File: SLR_Assignment_Calories_consumed.py
# -*- coding: utf-8 -*-
"""
Created on Tue Apr 28 15:27:00 2020

@author: RAVI
"""

# For reading data set
# importing necessary libraries
import pandas as pd # deals with data frame  
import numpy as np  # deals with numerical values

CC = pd.read_csv("C:/RAVI/Data science/Assignments/Module 6 Simple linear regression/DataSets/calories_consumed.csv")

CC.columns="Weightgained","Caloriesconsumed"
CC
import matplotlib.pylab as plt #for different types of plots

plt.scatter(x=CC['Caloriesconsumed'], y=CC['Weightgained'],color='green')# Scatter plot

np.corrcoef(CC.Caloriesconsumed, CC.Weightgained) #correlation

help(np.corrcoef)

import statsmodels.formula.api as smf
plt.hist(CC["Weightgained"])
plt.hist(CC["Caloriesconsumed"])
model = smf.ols('Weightgained ~ Caloriesconsumed', data=CC).fit()
model.summary()

#values prediction
#Confidence interval Calculation
pred1 = model.predict(pd.DataFrame(CC['Caloriesconsumed']))
pred1
print (model.conf_int(0.95)) # 95% confidence interval

res = CC.Weightgained - pred1
sqres = res*res
mse = np.mean(sqres)
rmse = np.sqrt(mse)

######### Model building on Transformed Data#############

# Log Transformation
# x = log(Caloriesconsumed); y = Weightgained
plt.scatter(x=np.log(CC['Caloriesconsumed']),y=CC['Weightgained'],color='brown')
np.corrcoef(np.log(CC.Caloriesconsumed), CC.Weightgained) #correlation

model2 = smf.ols('Weightgained ~ np.log(Caloriesconsumed)',data=CC).fit()
model2.summary()

pred2 = model2.predict(pd.DataFrame(CC['Caloriesconsumed']))
pred2
print(model2.conf_int(0.95)) # 95% confidence level

res2 = CC.Weightgained - pred2
sqres2 = res2*res2
mse2 = np.mean(sqres2)
rmse2 = np.sqrt(mse2)

# Exponential transformation
plt.scatter(x=CC['Caloriesconsumed'], y=np.log(CC['Weightgained']),color='orange')

np.corrcoef(CC.Caloriesconsumed, np.log(CC.Weightgained)) #correlation

model3 = smf.ols('np.log(Weightgained) ~ Caloriesconsumed',data=CC).fit()
model3.summary()

pred_log = model3.predict(pd.DataFrame(CC['Caloriesconsumed']))
pred_log
pred3 = np.exp(pred_log)
pred3
print(model3.conf_int(0.95)) # 95% confidence level

res3 = CC.Weightgained - pred3
sqres3 = res3*res3
mse3 = np.mean(sqres3)
rmse3 = np.sqrt(mse3)

############Polynomial model with 2 degree (quadratic model)  ;x = Caloriesconsumed*Caloriesconsumed; y = Weightgained############
#### input=x & X^2 (2-degree); output=y  ####
model4 = smf.ols('Weightgained ~ Caloriesconsumed+I(Caloriesconsumed*Caloriesconsumed)', data=CC).fit()
model4.summary()

pred_p2 = model4.predict(pd.DataFrame(CC['Caloriesconsumed']))
pred_p2

print(model3.conf_int(0.95)) # 95% confidence level

res4 = CC.Weightgained - pred_p2
sqres4 = res4*res4
mse4 = np.mean(sqres4)
rmse4 = np.sqrt(mse4)

###########Polynomial model with 3 degree (quadratic model)  ;x = Caloriesconsumed*Caloriesconsumed*Caloriesconsumed; y = Weightgained############
#### input=x & X^2 (2-degree); output=y  ####
model5 = smf.ols('Weightgained ~ Caloriesconsumed+I(Caloriesconsumed*Caloriesconsumed)+I(Caloriesconsumed*Caloriesconsumed*Caloriesconsumed)', data=CC).fit()
model5.summary()

pred_p3 = model5.predict(pd.DataFrame(CC['Caloriesconsumed']))
pred_p3

print(model5.conf_int(0.95)) # 95% confidence level

res5 = CC.Weightgained - pred_p3
sqres5 = res5*res5
mse5 = np.mean(sqres5)
rmse5 = np.sqrt(mse5)

File: SLR_Assignment_Calories_consumed.R
#  Simple Linear Regression Assignment #
#  1) Calories_consumed-> predict weight gained using calories consumed
#  Do the necessary transformations for input variables for getting better R^2 value for the model prepared.

library(readr)
library(ggplot2)
library(moments)
CC <- read_csv("C:/RAVI/Data science/Assignments/Module 6 Simple linear regression/DataSets/calories_consumed.csv")
View(CC)
attach(CC)
colnames(CC) <- c("Weightgained","Caloriesconsumed") 
View(CC)
summary(CC)
range(CC$Caloriesconsumed)
range(CC$Weightgained)
skewness(CC$Weightgained)
skewness(CC$Caloriesconsumed)

#Exploratory Data Analysis
boxplot(CC$Weightgained)
boxplot(CC$Caloriesconsumed)

#scatter plot for Caloriesconsumed vs Weightgained (Plot x,y)
plot(CC$Caloriesconsumed,CC$Weightgained)

#calculate correlation coefficient
cor(CC$Caloriesconsumed,CC$Weightgained)

#Simple Regression model
reg <- lm(CC$Weightgained~CC$Caloriesconsumed,data = CC)
summary(reg)

#values prediction
#Confidence interval Calculation
confint(reg,level = 0.95)
pred <- predict(reg,interval = "predict")
#predict function gives fit value and its lower and upeer values as a range
pred <- as.data.frame(pred)
pred

#####Plot Graph for both Actual values and also the predicted linear Graph(Actual:Red,Predicted:Blue)#########
ggplot() + 
  geom_point(aes(x =CC$Caloriesconsumed , y =CC$Weightgained ),
             colour='red') + 
  geom_line(aes(x = CC$Caloriesconsumed, y = predict(reg, newdata=CC)),
            colour='blue') + 
  ggtitle('Caloriesconsumed vs Weightgained') +
  xlab('Caloriesconsumed') +
  ylab('Weightgained')

cor(pred$fit,CC$Weightgained)

#Calculate Residuals "Errors"
reg$residuals
reg$residuals^2
mean(reg$residuals^2)
rmse <- sqrt(mean(reg$residuals^2))
rmse

############ Applying transformations##############
############ lOGORITHMIC MODEL    x = log(Caloriesconsumed); y = Weightgained ############
plot(log(CC$Caloriesconsumed),CC$Weightgained)
cor(log(CC$Caloriesconsumed),CC$Weightgained)

log_reg <- lm(CC$Weightgained ~ log(CC$Caloriesconsumed),data = CC)
summary(log_reg)

#values prediction
#Confidence interval Calculation
confint(log_reg,level = 0.95)
pred_log <- predict(log_reg,interval ="predict")
#predict function gives fit value and its lower and upeer values as a range
pred_log <- as.data.frame(pred_log)
pred_log


cor(pred_log$fit,CC$Weightgained)

rmse_log <- sqrt(mean(log_reg$residuals^2)) 
rmse_log
######or####
res_log=CC$Weightgained-pred_log$fit
rmse_log<-sqrt(mean(res_log^2))
rmse_log
##########Plot Graph for both Actual values and also the predicted linear Graph(Actual:Red,Predicted:Blue)#########
ggplot() + 
  geom_point(aes(x =CC$Caloriesconsumed , y =CC$Weightgained ),
             colour='red') + 
  geom_line(aes(x = CC$Caloriesconsumed, y = predict(log_reg, newdata=CC)),
            colour='blue') + 
  ggtitle('Caloriesconsumed vs Weightgained') +
  xlab('Caloriesconsumed') +
  ylab('Weightgained')

############ EXPONENTIAL MODEL   x = Caloriesconsumed; y = log(Weightgained) ############
plot(CC$Caloriesconsumed,log(CC$Weightgained))
cor(CC$Caloriesconsumed,log(CC$Weightgained))

log_reg2 <- lm(log(CC$Weightgained) ~ CC$Caloriesconsumed,data = CC)
summary(log_reg2)

#values prediction
#Confidence interval Calculation
confint(log_reg2,level = 0.95)
pred_log2 <- predict(log_reg2,interval ="predict")
#predict function gives fit value and its lower and upeer values as a range
pred_log2 <- as.data.frame(pred_log2)

log_reg2$residuals #output is log(AT) so we are getting less values apply antilog
pred<- exp(pred_log2)  #anti-log=exponential
pred

cor(pred$fit,CC$Weightgained)

res_log2=CC$Weightgained-pred$fit
rmse2 <- sqrt(mean(res_log2^2))
rmse2


##########Plot Graph for both Actual values and also the predicted linear Graph(Actual:Red,Predicted:Blue)#########
ggplot() + 
  geom_point(aes(x =CC$Caloriesconsumed , y =CC$Weightgained ),
             colour='red') + 
  geom_line(aes(x = CC$Caloriesconsumed, y = predict(log_reg2,data=CC)),
            colour='blue') + 
  ggtitle('Caloriesconsumed vs Weightgained') +
  xlab('Caloriesconsumed') +
  ylab('Weightgained')



############Polynomial model with 2 degree (quadratic model)  ;x =Caloriesconsumed^2 ; y = Weightgained ############
#### input=x & X^2 (2-degree); output=y  ####
reg_quad2<- lm(CC$Weightgained ~ CC$Caloriesconsumed+I(CC$Caloriesconsumed*CC$Caloriesconsumed),data =CC)
summary(reg_quad2)

#prediction
#Confidence interval Calculation
confint(reg_quad2,level = 0.95)
pred_quad2<-predict(reg_quad2,interval = "predict")
pred_quad2  <- as.data.frame(pred_quad2)
pred_quad2

resq=CC$Weightgained-pred_quad2$fit
rmse_quad<-sqrt(mean(resq^2))
rmse_quad

cor(pred_quad2$fit,CC$Weightgained)
##########Plot Graph for both Actual values and also the predicted linear Graph(Actual:Red,Predicted:Blue)#########
ggplot() + 
  geom_point(aes(x =CC$Caloriesconsumed , y =CC$Weightgained ),
             colour='red') + 
  geom_line(aes(x = CC$Caloriesconsumed, y = predict(reg_quad2,data=CC)),
            colour='blue') + 
  ggtitle('Caloriesconsumed vs Weightgained') +
  xlab('Caloriesconsumed') +
  ylab('Weightgained')

############Polynomial model with 3 degree (quadratic model)  ;x = Caloriesconsumed^3; y = Weightgained ############
#### input=x & X^2 & x^3 (3-degree); output=y  ####
reg_quad3<- lm(CC$Weightgained ~ CC$Caloriesconsumed+I(CC$Caloriesconsumed*CC$Caloriesconsumed)+I(CC$Caloriesconsumed*CC$Caloriesconsumed*CC$Caloriesconsumed),data =CC)
summary(reg_quad3)

#prediction
#Confidence interval Calculation
confint(reg_quad3,level = 0.95)
pred_quad3<-predict(reg_quad3,interval = "predict")
pred_quad3  <- as.data.frame(pred_quad3)
pred_quad3

cor(pred_quad3$fit,CC$Weightgained)

resq3=CC$Weightgained-pred_quad3$fit
rmse_quad3<-sqrt(mean(resq3^2))
rmse_quad3
##########Plot Graph for both Actual values and also the predicted linear Graph(Actual:Red,Predicted:Blue)#########
ggplot() + 
  geom_point(aes(x =CC$Caloriesconsumed , y =CC$Weightgained ),
             colour='red') + 
  geom_line(aes(x = CC$Caloriesconsumed, y = predict(reg_quad3,data=CC)),
            colour='blue') + 
  ggtitle('Caloriesconsumed vs Weightgained') +
  xlab('Caloriesconsumed') +
  ylab('Weightgained')








